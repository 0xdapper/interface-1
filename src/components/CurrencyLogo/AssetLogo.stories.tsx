import { ComponentMeta, ComponentStory } from '@storybook/react'
import React from 'react'
import { Box } from '../layout'
import { AssetLogoWithFallback } from './AssetLogo'

export default {
  title: 'WIP/Logos/AssetLogo',
  component: AssetLogoWithFallback,
  decorators: [
    (Story) => (
      <Box bg="background3" height={300} width={300}>
        <Story />
      </Box>
    ),
  ],
} as ComponentMeta<typeof AssetLogoWithFallback>

const Template: ComponentStory<typeof AssetLogoWithFallback> = (args) => (
  <AssetLogoWithFallback {...args} />
)

export const Primary = Template.bind({})
Primary.args = {
  url: 'https://token-icons.s3.amazonaws.com/eth.png',
  name: 'Maker',
}

export const Fallback = Template.bind({})
Fallback.args = {
  url: undefined,
  name: 'Maker',
}
