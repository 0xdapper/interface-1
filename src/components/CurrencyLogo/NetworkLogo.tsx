import React, { ComponentProps } from 'react'
import { Image } from 'react-native'
import { ETHEREUM_LOGO } from 'src/assets'
import { style } from 'src/components/CurrencyLogo/styles'
import { Box } from 'src/components/layout/Box'
import { ChainId, ChainIdTo } from 'src/constants/chains'

const LOGO_METADATA_BY_CHAIN: Partial<
  ChainIdTo<{
    logoSource: ComponentProps<typeof Image>['source'] | undefined
    // only used in testnets to help distinguish networks with same logo
    color: ComponentProps<typeof Box>['borderColor']
  }>
> = {
  [ChainId.MAINNET]: { logoSource: undefined, color: 'none' },
  [ChainId.RINKEBY]: {
    logoSource: ETHEREUM_LOGO,
    color: 'pink',
  },
  [ChainId.GOERLI]: {
    logoSource: ETHEREUM_LOGO,
    color: 'yellow',
  },
}

interface NetworkLogoProps {
  chainId: ChainId
  size?: number
}

export function NetworkLogo({ chainId, size = 10 }: NetworkLogoProps) {
  const { logoSource, color } = LOGO_METADATA_BY_CHAIN[chainId] || {}

  if (!logoSource) return null

  return (
    <Box borderWidth={2} borderColor={color} borderRadius="full">
      <Image
        style={[
          style.image,
          {
            width: size,
            height: size,
            borderRadius: size / 2,
          },
        ]}
        source={logoSource}
      />
    </Box>
  )
}
