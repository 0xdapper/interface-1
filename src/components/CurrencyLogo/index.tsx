import { Currency } from '@uniswap/sdk-core'
import React from 'react'
import { Image } from 'react-native'
import { SvgUri } from 'react-native-svg'
import { ETHEREUM_LOGO } from 'src/assets'
import { NetworkLogo } from 'src/components/CurrencyLogo/NetworkLogo'
import { style } from 'src/components/CurrencyLogo/styles'
import { getCurrencyLogoSrcs, maybeReplaceIPFSScheme } from 'src/components/CurrencyLogo/utils'
import { Box } from 'src/components/layout/Box'
import { ChainId } from 'src/constants/chains'

const DEFAULT_SIZE = 40

interface CurrencyLogoProps {
  currency: Currency
  size?: number
}

export function CurrencyLogo(props: CurrencyLogoProps) {
  const { size, currency } = props
  const currencyLogoSize = (size ?? DEFAULT_SIZE) - 4
  const networkSize = currencyLogoSize / 2.5
  return (
    <Box height={size} width={size}>
      <CurrencyLogoOnly currency={currency} size={currencyLogoSize} />
      {currency.chainId !== ChainId.MAINNET && (
        <Box bottom={0} position="absolute" right={0}>
          <NetworkLogo borderWidth={2} chainId={currency.chainId} size={networkSize} />
        </Box>
      )}
    </Box>
  )
}

function CurrencyLogoOnly({ currency, size = 40 }: CurrencyLogoProps) {
  const srcs: string[] = getCurrencyLogoSrcs(currency)

  if (currency?.isNative) {
    return (
      <Image
        source={ETHEREUM_LOGO}
        style={[style.image, { width: size, height: size, borderRadius: size / 2 }]}
      />
    )
  }

  // TODO(#95): Currently just uses the first URL in the source because unclear when we want to use a different one

  if (srcs.length > 0) {
    const src = srcs[0].toLowerCase()
    if (src.includes('.svg')) {
      return <SvgUri height={size} uri={srcs[0]} width={size} />
    }
    return (
      <Image
        source={srcs.map(maybeReplaceIPFSScheme)}
        style={[style.image, { width: size, height: size, borderRadius: size / 2 }]}
      />
    )
  }

  return null
}
