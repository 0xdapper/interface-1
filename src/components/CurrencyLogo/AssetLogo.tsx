import React from 'react'
import { Image, ImageStyle } from 'react-native'
import { Flex } from 'src/components/layout/Flex'
import { Text } from 'src/components/Text'

const tokenLogoStyle: ImageStyle = { width: 35, height: 35, borderRadius: 35 / 2 }

interface AssetLogoProps {
  url?: string
  // will fallback to name if no url
  name: string
}

export function AssetLogoWithFallback({ name, url }: AssetLogoProps) {
  return url ? (
    <Image source={{ uri: url }} style={tokenLogoStyle} />
  ) : (
    <Flex
      centered
      backgroundColor="backgroundContainer"
      borderRadius="full"
      height={tokenLogoStyle.height}
      width={tokenLogoStyle.width}>
      <Text color="textTertiary" variant="subHead2">
        {name[0]}
      </Text>
    </Flex>
  )
}
