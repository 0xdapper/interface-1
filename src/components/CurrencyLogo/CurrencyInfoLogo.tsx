import { ShadowProps } from '@shopify/restyle'
import React, { useState } from 'react'
import { Image } from 'react-native'
import { useAppTheme } from 'src/app/hooks'
import { NetworkLogo } from 'src/components/CurrencyLogo/NetworkLogo'
import { style } from 'src/components/CurrencyLogo/styles'
import { Box } from 'src/components/layout/Box'
import { Text } from 'src/components/Text'
import { ChainId } from 'src/constants/chains'
import { CurrencyInfo } from 'src/features/dataApi/types'
import { Theme } from 'src/styles/theme'

const DEFAULT_SIZE = 36
const NETWORK_LOGO_SIZE = 16
const SHADOW_OFFSET: ShadowProps<Theme>['shadowOffset'] = { width: 0, height: 2 }
const THIN_BORDER = 0.5

interface CurrencyInfoLogoProps {
  currencyInfo: CurrencyInfo
  size?: number
}

export function CurrencyInfoLogo({ currencyInfo, size = DEFAULT_SIZE }: CurrencyInfoLogoProps) {
  const { currency } = currencyInfo
  const notOnMainnet = currency.chainId !== ChainId.Mainnet

  return (
    <Box alignItems="center" height={size} justifyContent="center" width={size}>
      <CurrencyInfoLogoOnly currencyInfo={currencyInfo} size={size} />
      {notOnMainnet && (
        <Box
          bottom={0}
          position="absolute"
          right={0}
          shadowColor="black"
          shadowOffset={SHADOW_OFFSET}
          shadowOpacity={0.1}
          shadowRadius={2}>
          <NetworkLogo chainId={currency.chainId} size={NETWORK_LOGO_SIZE} />
        </Box>
      )}
    </Box>
  )
}

export function CurrencyInfoLogoOnly({ currencyInfo, size = 40 }: CurrencyInfoLogoProps) {
  const { logoUrl, currency } = currencyInfo

  const [imageError, setImageError] = useState(false)

  const onImageNotFound = () => {
    setImageError(true)
  }

  const theme = useAppTheme()

  if (logoUrl && !imageError) {
    return (
      <Image
        source={{ uri: logoUrl }}
        style={[
          style.image,
          {
            backgroundColor: theme.colors.textTertiary,
            width: size,
            height: size,
            borderRadius: size / 2,
            borderColor: theme.colors.backgroundOutline,
            borderWidth: THIN_BORDER,
          },
        ]}
        onError={() => onImageNotFound()}
      />
    )
  } else {
    return (
      <Box
        alignItems="center"
        bg="backgroundAction"
        height={size}
        justifyContent="center"
        px="xxs"
        style={{ borderRadius: size / 2 }}
        width={size}>
        <Text adjustsFontSizeToFit color="textSecondary" numberOfLines={1} textAlign="center">
          {currency.symbol?.slice(0, 5).toUpperCase()}
        </Text>
      </Box>
    )
  }
}
