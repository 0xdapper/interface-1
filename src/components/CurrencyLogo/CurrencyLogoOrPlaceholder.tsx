import { Currency } from '@uniswap/sdk-core'
import React from 'react'
import { CurrencyLogoOnly } from 'src/components/CurrencyLogo'
import { Box } from 'src/components/layout'

export function CurrencyLogoOrPlaceholder(props: { currency: Nullable<Currency>; size: number }) {
  const { currency, size } = props
  if (!currency) {
    return (
      <Box backgroundColor="backgroundContainer" borderRadius="full" height={size} width={size} />
    )
  }

  return <CurrencyLogoOnly currency={currency} size={size} />
}
