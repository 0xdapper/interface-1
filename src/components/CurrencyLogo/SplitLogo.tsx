import React from 'react'
import { CurrencyLogo } from 'src/components/CurrencyLogo'
import { Box } from 'src/components/layout/Box'
import { CurrencyInfo } from 'src/features/dataApi/types'

interface Props {
  inputCurrencyInfo: NullUndefined<CurrencyInfo>
  outputCurrencyInfo: NullUndefined<CurrencyInfo>
  size: number
}

/*
 * Logo, where left 50% of width is taken from one icon (its left 50%)
 * and right side is taken from another icon (its right 50%)
 */
export function SplitLogo({ size, inputCurrencyInfo, outputCurrencyInfo }: Props): JSX.Element {
  const iconSize = size / 2
  return (
    <Box height={size} width={size}>
      <Box
        left={0}
        overflow="hidden"
        position="absolute"
        top={0}
        width={iconSize - 1 /* -1 to allow for space between the icons */}>
        <CurrencyLogo hideNetworkLogo currencyInfo={inputCurrencyInfo} size={size} />
      </Box>
      <Box
        flexDirection="row-reverse"
        overflow="hidden"
        position="absolute"
        right={0}
        top={0}
        width={iconSize - 1 /* -1 to allow for space between the icons */}>
        <CurrencyLogo hideNetworkLogo currencyInfo={outputCurrencyInfo} size={size} />
      </Box>
    </Box>
  )
}
