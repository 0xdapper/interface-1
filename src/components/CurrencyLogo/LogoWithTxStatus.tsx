import { Currency } from '@uniswap/sdk-core'
import React from 'react'
import { useAppTheme } from 'src/app/hooks'
import AlertTriangle from 'src/assets/icons/alert-triangle.svg'
import Approve from 'src/assets/icons/approve.svg'
import IncomingArrow from 'src/assets/icons/arrow-down-in-circle.svg'
import OutgoingArrow from 'src/assets/icons/arrow-up-in-circle.svg'
import UnknownStatus from 'src/assets/icons/question-in-circle.svg'
import SlashCircleIcon from 'src/assets/icons/slash-circle.svg'
import { CurrencyLogoOrPlaceholder } from 'src/components/CurrencyLogo/CurrencyLogoOrPlaceholder'
import { NFTViewer } from 'src/components/images/NFTViewer'
import { Box } from 'src/components/layout/Box'
import { AssetType } from 'src/entities/assets'
import { TXN_HISTORY_SIZING } from 'src/features/transactions/SummaryCards/TransactionSummaryLayout'
import { NFTTradeType, TransactionStatus, TransactionType } from 'src/features/transactions/types'
import { logger } from 'src/utils/logger'

interface LogoWithTxStatusProps {
  assetType: AssetType
  txType: TransactionType
  txStatus: TransactionStatus
  size: {
    primaryImage: number
    secondaryImage: number
  }
}

interface CurrencyStatusProps extends LogoWithTxStatusProps {
  assetType: AssetType.Currency
  currency?: Currency | null
}

interface NFTStatusProps extends LogoWithTxStatusProps {
  assetType: AssetType.ERC721 | AssetType.ERC1155
  nftImageUrl?: string
  nftTradeType?: NFTTradeType
}

export function LogoWithTxStatus(props: CurrencyStatusProps | NFTStatusProps) {
  const { assetType, txType, txStatus, size } = props
  const theme = useAppTheme()

  const logo =
    assetType === AssetType.Currency ? (
      <CurrencyLogoOrPlaceholder currency={props.currency} size={size.primaryImage} />
    ) : (
      <NFTLogoOrPlaceholder nftImageUrl={props.nftImageUrl} size={size.primaryImage} />
    )

  const fill = theme.colors.backgroundBackdrop
  const gray = theme.colors.textSecondary
  const green = theme.colors.accentSuccess
  const yellow = theme.colors.accentWarning
  const statusSize = size.secondaryImage

  const getTxStatusIcon = () => {
    if (txStatus === TransactionStatus.Failed) {
      return <AlertTriangle color={yellow} fill={fill} height={statusSize} width={statusSize} />
    }
    if (txStatus === TransactionStatus.Cancelled) {
      return (
        <SlashCircleIcon
          color={theme.colors.backgroundOutline}
          fill={theme.colors.black}
          fillOpacity={1}
          height={statusSize}
          width={statusSize}
        />
      )
    }
    switch (txType) {
      case TransactionType.Approve:
        return <Approve color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.NFTApprove:
        return <Approve color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.Send:
        return <OutgoingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.Receive:
        return <IncomingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.NFTTrade:
        if (assetType === AssetType.ERC721 && props.nftTradeType === NFTTradeType.SELL) {
          return <OutgoingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
        }
        return <IncomingArrow color={green} fill={fill} height={statusSize} width={statusSize} />
      case TransactionType.Unknown:
        return <UnknownStatus color={gray} fill={fill} height={statusSize} width={statusSize} />
    }
    logger.info(
      'statusIcon',
      'GenerateStatusIcon',
      'Could not find icon for transaction type:',
      txType
    )
    return null
  }
  const statusIcon = getTxStatusIcon()
  return (
    <>
      {logo}
      <Box bottom={1} position="absolute" right={1}>
        {statusIcon}
      </Box>
    </>
  )
}

function NFTLogoOrPlaceholder(props: { nftImageUrl?: string; size: number }) {
  const { nftImageUrl, size } = props
  return (
    <Box
      alignItems="center"
      backgroundColor="backgroundContainer"
      borderRadius="xs"
      height={size}
      justifyContent="center"
      overflow="hidden"
      width={size}>
      {nftImageUrl && <NFTViewer uri={nftImageUrl} />}
    </Box>
  )
}

/**
 * Swap icons lockup, fall back to single icon plus warning if in failed state.
 */
export function DoubleCurrencyLogoWithTxStatus({
  currency,
  otherCurrency,
  status,
  showCancelIcon,
}: {
  currency: NullUndefined<Currency>
  otherCurrency: NullUndefined<Currency>
  status: TransactionStatus
  showCancelIcon: boolean
}) {
  if (
    status === TransactionStatus.Failed ||
    (showCancelIcon && status === TransactionStatus.Cancelled)
  ) {
    return (
      <LogoWithTxStatus
        assetType={AssetType.Currency}
        currency={currency}
        size={TXN_HISTORY_SIZING}
        txStatus={status}
        txType={TransactionType.Swap}
      />
    )
  }
  return (
    <>
      <Box left={2} position="absolute" top={2}>
        <CurrencyLogoOrPlaceholder
          currency={otherCurrency}
          size={TXN_HISTORY_SIZING.primaryImage}
        />
      </Box>
      <Box bottom={0} position="absolute" right={0}>
        <CurrencyLogoOrPlaceholder currency={currency} size={TXN_HISTORY_SIZING.primaryImage} />
      </Box>
    </>
  )
}
