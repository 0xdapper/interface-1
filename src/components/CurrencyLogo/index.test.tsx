import { render } from '@testing-library/react-native'
import React from 'react'
import { CurrencyLogo } from 'src/components/CurrencyLogo'
import { ChainId } from 'src/constants/chains'
import { UNI } from 'src/constants/tokens'
import { renderWithTheme } from 'src/test/render'

jest.mock('src/assets', () => 'ethereum-logo.png')

it('renders a currency logo without network logo', () => {
  const tree = render(<CurrencyLogo currency={UNI[ChainId.Mainnet]} size={20} />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders a currency logo with network logo', () => {
  const tree = renderWithTheme(<CurrencyLogo currency={UNI[ChainId.Goerli]} size={20} />)
  expect(tree).toMatchSnapshot()
})
