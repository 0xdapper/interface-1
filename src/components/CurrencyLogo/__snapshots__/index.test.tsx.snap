// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a currency logo with network logo 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "height": 20,
        "width": 20,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "bottom": 0,
          "display": "flex",
          "position": "absolute",
          "right": 0,
        },
      ]
    }
  />
</View>
`;

exports[`renders a currency logo without network logo 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
        "height": 20,
        "width": 20,
      },
    ]
  }
>
  <Image
    source={
      Array [
        Object {
          "uri": "https://raw.githubusercontent.com/uniswap/assets/master/blockchains/ethereum/assets/0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984/logo.png",
        },
      ]
    }
    style={
      Array [
        Object {
          "backgroundColor": "slategrey",
          "resizeMode": "cover",
        },
        Object {
          "borderRadius": 8,
          "height": 16,
          "width": 16,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "bottom": 0,
          "display": "flex",
          "position": "absolute",
          "right": 0,
        },
      ]
    }
  />
</View>
`;
